flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoderDelegateTest.java

```diff
@@ -0,0 +1,324 @@
+		}
+
+		return serializedBuffers;
+	}
+
+	private int[] getBufferSizes(ByteBuf[] buffers) {
+		int[] sizes = new int[buffers.length];
+		for (int i = 0; i < sizes.length; ++i) {
+			sizes[i] = buffers[i].readableBytes();
+		}
+
+		return sizes;
+	}
+
+	private ByteBuf mergeBuffers(ByteBuf[] buffers) {
+		ByteBuf allData = ALLOCATOR.buffer();
+		for (ByteBuf buffer : buffers) {
+			allData.writeBytes(buffer);
+		}
+
+		return allData;
+	}
+
+	private List<ByteBuf> partitionBuffer(ByteBuf buffer, int[] sizes) {
+		// For each group of three messages [A, B, C], we repartition them as three buffers:
+		//  1. First half of A.
+		//  2. Second half of A + B + first half of C.
+		//  3. Second half of C.
+		List<Integer> partitionPositions = new ArrayList<>();
+
+		int position = 0;
+		partitionPositions.add(position);
+
+		for (int i = 0; i < sizes.length / 3 * 3; i += 3) {
+			position += sizes[i] / 2;
+			partitionPositions.add(position);
+
+			position += (sizes[i] / 2 + sizes[i + 1] + sizes[i + 2] / 2);
+			partitionPositions.add(position);
+
+			position += sizes[i + 2] / 2;
+			partitionPositions.add(position);
+		}
+
+		// Partitions the remaining data into three parts averagely.
+		int remainingBytes = buffer.readableBytes() - position;
+
+		position += (remainingBytes / 3);
+		partitionPositions.add(position);
+
+		position += (remainingBytes / 3 * 2);
+		partitionPositions.add(position);
+
+		partitionPositions.add(buffer.readableBytes());
+
+		// Partitions the buffers according to the selected partition positions.
+		List<ByteBuf> result = new ArrayList<>();
+		for (int i = 1; i < partitionPositions.size(); ++i) {
+			ByteBuf partitionedBuffer = ALLOCATOR.buffer();
+			partitionedBuffer.writeBytes(
+				buffer,
+				partitionPositions.get(i - 1),
+				partitionPositions.get(i) - partitionPositions.get(i - 1));
+			result.add(partitionedBuffer);
+		}
+
+		return result;
+	}
+
+	private List<NettyMessage> decodedMessages(EmbeddedChannel channel, List<ByteBuf> inputBuffers) {
+		for (ByteBuf buffer : inputBuffers) {
+			channel.writeInbound(buffer);
+		}
+
+		channel.runPendingTasks();
+
+		List<NettyMessage> decodedMessages = new ArrayList<>();
+		Object input;
+		while ((input = channel.readInbound()) != null) {
+			assertTrue(input instanceof NettyMessage);
+			decodedMessages.add((NettyMessage) input);
+		}
+
+		return decodedMessages;
+	}
+
+	private void verifyDecodedMessages(List<NettyMessage> expected, List<NettyMessage> decodedMessages) {
+		assertEquals(expected.size(), decodedMessages.size());
+		for (int i = 0; i < expected.size(); ++i) {
+			assertEquals(expected.get(i).getClass(), decodedMessages.get(i).getClass());
+
+			if (expected.get(i) instanceof NettyMessage.BufferResponse) {
+				BufferResponse expectedBufferResponse = (BufferResponse) expected.get(i);
+				BufferResponse decodedBufferResponse = (BufferResponse) decodedMessages.get(i);
+
+				assertEquals(expectedBufferResponse.backlog, decodedBufferResponse.backlog);
```

Comment:
Extract a static method for verifying the `BufferResponse` which can be reused for other class tests.

```
static verifyBufferResponse(BufferResponse original, BufferResponse expected) {
        assertEquals(expected.backlog, original.backlog);
        assertEquals(expected.sequenceNumber, original.sequenceNumber);
	assertEquals(expected.bufferSize, original.bufferSize);
	assertEquals(expected.receiverId, original.receiverId);
}
```

Reply:
Extractor the comparison of buffer response header.